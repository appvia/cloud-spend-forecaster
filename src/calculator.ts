import { Component, ComponentOutput, ComputeNode, Interval } from '../module'

export function putComponentsIntoIntervals(components: Component[]) {
  return (interval: Interval): Interval => {
    return { ...interval, components }
  }
}

export function putNodesIntoIntervals(node: ComputeNode) {
  return (interval: Interval): Interval => {
    return { ...interval, nodes: node }
  }
}

export function calculateResourceNeededForRequestsInterval(interval: Interval): Interval {
  const intervalComponents = interval.components.map((component: ComponentOutput): ComponentOutput => {
    return {
      ...component,
      needCpuForRequests: interval.requests * component.requestToCpu,
      needMemoryForRequests: interval.requests * component.requestToMemory,
    }
  })
  return {
    ...interval,
    components: intervalComponents,
  }
}

export function calculateReplicasNeededForInterval(interval: Interval): Interval {
  const intervalComponents = interval.components.map((component: ComponentOutput): ComponentOutput => {
    const needCpuReplica = Math.ceil(component.needCpuForRequests / (component.limitCpu - component.baselineCpu))
    const needMemoryReplica = Math.ceil(
      component.needMemoryForRequests / (component.limitMemory - component.baselineMemory),
    )

    return {
      ...component,
      needCpuReplica,
      needMemoryReplica,
      needReplica: Math.max(needMemoryReplica, needCpuReplica, component.minReplica),
    }
  })
  return {
    ...interval,
    components: intervalComponents,
    needPods: intervalComponents.reduce((accumulator, component) => accumulator + component.needReplica, 0),
  }
}

export function calculateResourceNeededForInterval(interval: Interval): Interval {
  const intervalComponents = interval.components.map((component: ComponentOutput): ComponentOutput => {
    return {
      ...component,
      needCpu: component.needReplica * component.baselineCpu + component.needCpuForRequests,
      needMemory: component.needReplica * component.baselineMemory + component.needMemoryForRequests,
    }
  })
  return {
    ...interval,
    components: intervalComponents,
    needCpu: intervalComponents.reduce((accumulator, component) => accumulator + component.needCpu, 0),
    needMemory: intervalComponents.reduce((accumulator, component) => accumulator + component.needMemory, 0),
  }
}

export function calculateNodesNeededForInterval(interval: Interval): Interval {
  const needNodesByCpu = Math.ceil(interval.needCpu / interval.nodes.availableCpu)
  const needNodesByMemory = Math.ceil(interval.needMemory / interval.nodes.availableMemory)
  const needNodesByPods = Math.ceil(interval.needPods / interval.nodes.maxPods)

  return { ...interval, needNodesByCpu, needNodesByMemory, needNodesByPods }
}

export function calcCapacity(replica: number, limit: number, to_request: number, baseline: number): number {
  return Math.floor(replica * (limit * to_request - baseline))
}

export function withinRange(min: number, max: number, number: number): number {
  return Math.ceil(Math.max(min, Math.min(max, number)))
}

export function calculateDesiredResourceForInterval(interval: Interval): Interval {
  const intervalComponents = interval.components.map((component: ComponentOutput): ComponentOutput => {
    const desiredReplica = withinRange(
      component.minReplica,
      component.maxReplica,
      (100 / component.scalingThresholdCpu) * component.needCpuReplica,
    )
    const desiredCpu = desiredReplica * component.limitCpu
    const desiredMemory = desiredReplica * component.limitMemory
    return {
      ...component,
      desiredReplica,
      desiredCpu,
      desiredMemory,
    }
  })
  return {
    ...interval,
    components: intervalComponents,
    desiredPods: intervalComponents.reduce(
      (accumulator: number, component: ComponentOutput): number => accumulator + component.desiredReplica,
      0,
    ),
    desiredCpu: intervalComponents.reduce(
      (accumulator: number, component: ComponentOutput): number => accumulator + component.desiredCpu,
      0,
    ),
    desiredMemory: intervalComponents.reduce(
      (accumulator: number, component: ComponentOutput): number => accumulator + component.desiredMemory,
      0,
    ),
  }
}

export function calculateNodesDesiredForInterval(interval: Interval): Interval {
  const desiredNodesByCpu = Math.ceil(interval.desiredCpu / interval.nodes.availableCpu)
  const desiredNodesByMemory = Math.ceil(interval.desiredMemory / interval.nodes.availableMemory)
  const desiredNodesByPods = Math.ceil(interval.desiredPods / interval.nodes.maxPods)

  return {
    ...interval,
    desiredNodesByCpu,
    desiredNodesByMemory,
    desiredNodesByPods,
    desiredNodes: withinRange(
      interval.nodes.minNodes,
      interval.nodes.maxNodes,
      Math.max(desiredNodesByCpu, desiredNodesByMemory, desiredNodesByPods),
    ),
  }
}

export function calculateReadyNodes(interval: Interval, index: number, intervals: Interval[]): Interval {
  return {
    ...interval,
    readyNodes:
      index > interval.nodes.scalingIntervals
        ? intervals[index + 1 - interval.nodes.scalingIntervals].desiredNodes
        : interval.nodes.minNodes,
  }
}

export function calculatePendingPods(interval: Interval, index: number, intervals: Interval[]): Interval {
  const intervalComponents = interval.components.map((component: ComponentOutput, cid: number) => {
    let pendingReplica = component.minReplica

    if (index + 1 - component.scalingIntervals > 0) {
      pendingReplica = intervals[index + 1 - component.scalingIntervals].components[cid].desiredReplica
    }

    return {
      ...component,
      pendingReplica,
    }
  })

  return { ...interval, components: intervalComponents }
}

export function calculateReadyPods(interval: Interval): Interval {
  const intervalComponents = interval.components.map((component: ComponentOutput): ComponentOutput => {
    const readyReplica = component.pendingReplica
    return {
      ...component,
      readyReplica,
    }
  })

  let cid = 0
  do {
    intervalComponents[cid].readyReplica = Math.max(
      intervalComponents[cid].readyReplica - 1,
      intervalComponents[cid].minReplica,
    )
    cid = cid === intervalComponents.length - 1 ? 0 : cid + 1
  } while (calculateIfOverSubscribed(intervalComponents, interval.nodes, interval.readyNodes))

  return { ...interval, components: intervalComponents }
}

export function calculateIfOverSubscribed(
  components: ComponentOutput[],
  nodes: ComputeNode,
  readyNodes: number,
): boolean {
  const pendingPodCount = components.reduce(
    (accumulator: number, component: ComponentOutput): number => accumulator + component.readyReplica,
    0,
  )
  if (pendingPodCount > readyNodes * nodes.maxPods) {
    return true
  }
  const pendingCpu = components.reduce(
    (accumulator: number, component: ComponentOutput): number =>
      accumulator + component.readyReplica * component.limitCpu,
    0,
  )

  if (pendingCpu > readyNodes * nodes.availableCpu) {
    return true
  }

  const pendingMemory = components.reduce(
    (accumulator: number, component: ComponentOutput): number =>
      accumulator + component.readyReplica * component.limitMemory,
    0,
  )
  if (pendingMemory > readyNodes * nodes.availableMemory) {
    return true
  }

  return false
}

export function calculateCosts(interval: Interval): Interval {
  return { ...interval, cost: interval.desiredNodes * interval.nodes.cost }
}

export function calculatePenalties(failedRequestPenalty: number) {
  return (interval: Interval): Interval => {
    return {
      ...interval,
      failedRequestPenalty: interval.failedRequests * failedRequestPenalty,
    }
  }
}

export function calculateReadyRequestCapacity(interval: Interval): Interval {
  const intervalComponents = interval.components.map((component) => {
    const cpuCapacity = (component.readyReplica * (component.limitCpu - component.baselineCpu)) / component.requestToCpu

    const memoryCapacity =
      (component.readyReplica * (component.limitMemory - component.baselineMemory)) / component.requestToMemory

    const readyRequestCapacity = Math.floor(Math.min(cpuCapacity, memoryCapacity))
    return { ...component, readyRequestCapacity }
  })
  return {
    ...interval,
    components: intervalComponents,
    readyRequestCapacity: Math.min(...intervalComponents.map((component) => component.readyRequestCapacity)),
  }
}

export function calculateFailedRequests(interval: Interval): Interval {
  const intervalComponents = interval.components.map((component) => {
    return {
      ...component,
      failedRequests: Math.ceil(Math.max(0, interval.requests - component.readyRequestCapacity)),
    }
  })
  return {
    ...interval,
    components: intervalComponents,
    failedRequests: Math.max.apply(
      null,
      intervalComponents.map((component) => component.failedRequests),
    ),
  }
}
