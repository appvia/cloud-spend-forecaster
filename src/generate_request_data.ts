export interface Puller {
  time: number
  requests: number
  desc?: String
}

export interface RequestData {
  time: number
  requests: number
}

export function generate (pullers: Puller[], units: number): RequestData[] {
  const data = []
  pullers.forEach((from: Puller, i: number) => {
    if (i + 1 == pullers.length) return
    const next: Puller = pullers[i + 1]

    const toCreate = (next.time - from.time) / units
    for (let i = 0; i < toCreate; i++) {
      const foo =
        ((next.requests - from.requests) / toCreate) * i + from.requests
      data.push({
        time: i * units + from.time,
        requests: Math.ceil(foo)
      })
    }
  })
  return data
}
