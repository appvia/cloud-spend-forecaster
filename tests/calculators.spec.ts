import * as mod from '../src/calculator'

describe('calculator', () => {
  it('calculateResourceNeededForRequestsInterval', () =>
    expect(
      mod.calculateResourceNeededForRequestsInterval({
        requests: 100,
        components: [{ requestToCpu: 3, requestToMemory: 4 }]
      })
    ).toMatchSnapshot())

  it('calculateReplicasNeededForInterval', () =>
    expect(
      mod.calculateReplicasNeededForInterval({
        requests: 100,
        components: [
          {
            needCpuForRequests: 30,
            limitCpu: 40,
            baselineCpu: 4,
            minReplica: 1,
            needMemoryForRequests: 20,
            limitMemory: 200,
            baselineMemory: 500
          }
        ]
      })
    ).toMatchSnapshot())

  it('calculateResourceNeededForInterval', () =>
    expect(
      mod.calculateResourceNeededForInterval({
        components: [
          {
            needReplica: 2,
            baselineCpu: 2,
            needCpuForRequests: 200,
            baselineMemory: 10,
            needMemoryForRequests: 15
          }
        ]
      })
    ).toMatchSnapshot())

  it('calculateNodesNeededForInterval', () =>
    expect(
      mod.calculateNodesNeededForInterval({
        needCpu: 10000,
        needMemory: 200000,
        needPods: 1000,
        nodes: {
          availableCpu: 100,
          availableMemory: 200,
          maxPods: 100
        }
      })
    ).toMatchSnapshot())

  it('calculateNodesDesiredForInterval', () =>
    expect(
      mod.calculateNodesDesiredForInterval({
        desiredCpu: 10000,
        desiredMemory: 200000,
        desiredPods: 1000,
        nodes: {
          availableCpu: 100,
          availableMemory: 200,
          maxPods: 100
        }
      })
    ).toMatchSnapshot())

  it('calculateDesiredResourceForInterval', () =>
    expect(
      mod.calculateDesiredResourceForInterval({
        components: [
          {
            minReplica: 2,
            maxReplica: 200,
            scalingThresholdCpu: 70,
            needCpuReplica: 10,
            limitCpu: 100,
            limitMemory: 200
          }
        ]
      })
    ).toMatchSnapshot())

  it('calculateReadyNodes', () => {
    const arr = [
      { desiredNodes: 10, nodes: { scalingIntervals: 3 } },
      { desiredNodes: 10, nodes: { scalingIntervals: 3 } },
      { desiredNodes: 10, nodes: { scalingIntervals: 3 } },
      { desiredNodes: 5, nodes: { scalingIntervals: 3 } },
      { desiredNodes: 10, nodes: { scalingIntervals: 3 } },
      { desiredNodes: 10, nodes: { scalingIntervals: 3 } },
      { desiredNodes: 10, nodes: { scalingIntervals: 3 } }
    ]
    return expect(arr.map(mod.calculateReadyNodes)).toMatchSnapshot()
  })

  it('calculatePendingPods', () => {
    const arr = [
      {
        components: [{ scalingIntervals: 2, desiredReplica: 3, minReplica: 1 }]
      },
      {
        components: [{ scalingIntervals: 2, desiredReplica: 3, minReplica: 1 }]
      },
      {
        components: [{ scalingIntervals: 2, desiredReplica: 3, minReplica: 1 }]
      },
      {
        components: [{ scalingIntervals: 2, desiredReplica: 3, minReplica: 1 }]
      },
      {
        components: [{ scalingIntervals: 2, desiredReplica: 3, minReplica: 1 }]
      },
      {
        components: [{ scalingIntervals: 2, desiredReplica: 10, minReplica: 1 }]
      },
      {
        components: [{ scalingIntervals: 2, desiredReplica: 4, minReplica: 1 }]
      },
      {
        components: [{ scalingIntervals: 2, desiredReplica: 3, minReplica: 1 }]
      },
      {
        components: [{ scalingIntervals: 2, desiredReplica: 3, minReplica: 1 }]
      }
    ]
    return expect(arr.map(mod.calculatePendingPods)).toMatchSnapshot()
  })

  test.each([
    {
      components: [{ readyReplica: 100 }],
      nodes: { maxPods: 10 },
      readyNodes: 2,
      testing: 'pods',
      expected: true
    },
    {
      components: [{ readyReplica: 20, limitCpu: 1000 }],
      nodes: { maxPods: 100, availableCpu: 1 },
      readyNodes: 1,
      testing: 'cpu',
      expected: true
    },
    {
      components: [{ readyReplica: 20, limitCpu: 1, limitMemory: 100 }],
      nodes: { maxPods: 100, availableCpu: 1000, availableMemory: 1 },
      readyNodes: 1,
      testing: 'memory',
      expected: true
    },
    {
      components: [{ readyReplica: 2, limitCpu: 1, limitMemory: 1 }],
      nodes: { maxPods: 100, availableCpu: 1000, availableMemory: 100 },
      readyNodes: 1,
      testing: 'ok',
      expected: false
    }
  ])(
    'calculateIfOverSubscribed, $testing',
    ({ components, nodes, readyNodes, testing, expected }) =>
      expect(mod.calculateIfOverSubscribed(components, nodes, readyNodes)).toBe(
        expected
      )
  )

  it.todo('calculatePenalties')
  it.todo('calculateReadyRequestCapacity')
  it.todo('calculateFailedRequests')
  it.todo('calculateReadyPods')
  it.todo('calcCapacity')
})
