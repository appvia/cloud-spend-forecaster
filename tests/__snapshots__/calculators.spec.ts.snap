// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculator calculateDesiredResourceForInterval 1`] = `
Object {
  "components": Array [
    Object {
      "desiredCpu": 1500,
      "desiredMemory": 3000,
      "desiredReplica": 15,
      "limitCpu": 100,
      "limitMemory": 200,
      "maxReplica": 200,
      "minReplica": 2,
      "needCpuReplica": 10,
      "scalingThresholdCpu": 70,
    },
  ],
  "desiredCpu": 1500,
  "desiredMemory": 3000,
  "desiredPods": 15,
}
`;

exports[`calculator calculateNodesDesiredForInterval 1`] = `
Object {
  "desiredCpu": 10000,
  "desiredMemory": 200000,
  "desiredNodes": 1000,
  "desiredNodesByCpu": 100,
  "desiredNodesByMemory": 1000,
  "desiredNodesByPods": 10,
  "desiredPods": 1000,
  "nodes": Object {
    "availableCpu": 100,
    "availableMemory": 200,
    "maxPods": 100,
  },
}
`;

exports[`calculator calculateNodesNeededForInterval 1`] = `
Object {
  "needCpu": 10000,
  "needMemory": 200000,
  "needNodesByCpu": 100,
  "needNodesByMemory": 1000,
  "needNodesByPods": 10,
  "needPods": 1000,
  "nodes": Object {
    "availableCpu": 100,
    "availableMemory": 200,
    "maxPods": 100,
  },
}
`;

exports[`calculator calculatePendingPods 1`] = `
Array [
  Object {
    "components": Array [
      Object {
        "desiredReplica": 3,
        "minReplica": 1,
        "pendingReplica": 1,
        "scalingIntervals": 2,
      },
    ],
  },
  Object {
    "components": Array [
      Object {
        "desiredReplica": 3,
        "minReplica": 1,
        "pendingReplica": 1,
        "scalingIntervals": 2,
      },
    ],
  },
  Object {
    "components": Array [
      Object {
        "desiredReplica": 3,
        "minReplica": 1,
        "pendingReplica": 3,
        "scalingIntervals": 2,
      },
    ],
  },
  Object {
    "components": Array [
      Object {
        "desiredReplica": 3,
        "minReplica": 1,
        "pendingReplica": 3,
        "scalingIntervals": 2,
      },
    ],
  },
  Object {
    "components": Array [
      Object {
        "desiredReplica": 3,
        "minReplica": 1,
        "pendingReplica": 3,
        "scalingIntervals": 2,
      },
    ],
  },
  Object {
    "components": Array [
      Object {
        "desiredReplica": 10,
        "minReplica": 1,
        "pendingReplica": 3,
        "scalingIntervals": 2,
      },
    ],
  },
  Object {
    "components": Array [
      Object {
        "desiredReplica": 4,
        "minReplica": 1,
        "pendingReplica": 10,
        "scalingIntervals": 2,
      },
    ],
  },
  Object {
    "components": Array [
      Object {
        "desiredReplica": 3,
        "minReplica": 1,
        "pendingReplica": 4,
        "scalingIntervals": 2,
      },
    ],
  },
  Object {
    "components": Array [
      Object {
        "desiredReplica": 3,
        "minReplica": 1,
        "pendingReplica": 3,
        "scalingIntervals": 2,
      },
    ],
  },
]
`;

exports[`calculator calculateReadyNodes 1`] = `
Array [
  Object {
    "desiredNodes": 10,
    "nodes": Object {
      "scalingIntervals": 3,
    },
    "readyNodes": undefined,
  },
  Object {
    "desiredNodes": 10,
    "nodes": Object {
      "scalingIntervals": 3,
    },
    "readyNodes": undefined,
  },
  Object {
    "desiredNodes": 10,
    "nodes": Object {
      "scalingIntervals": 3,
    },
    "readyNodes": undefined,
  },
  Object {
    "desiredNodes": 5,
    "nodes": Object {
      "scalingIntervals": 3,
    },
    "readyNodes": undefined,
  },
  Object {
    "desiredNodes": 10,
    "nodes": Object {
      "scalingIntervals": 3,
    },
    "readyNodes": 10,
  },
  Object {
    "desiredNodes": 10,
    "nodes": Object {
      "scalingIntervals": 3,
    },
    "readyNodes": 5,
  },
  Object {
    "desiredNodes": 10,
    "nodes": Object {
      "scalingIntervals": 3,
    },
    "readyNodes": 10,
  },
]
`;

exports[`calculator calculateReplicasNeededForInterval 1`] = `
Object {
  "components": Array [
    Object {
      "baselineCpu": 4,
      "baselineMemory": 500,
      "limitCpu": 40,
      "limitMemory": 200,
      "minReplica": 1,
      "needCpuForRequests": 30,
      "needCpuReplica": 1,
      "needMemoryForRequests": 20,
      "needMemoryReplica": -0,
      "needReplica": 1,
    },
  ],
  "needPods": 1,
  "requests": 100,
}
`;

exports[`calculator calculateResourceNeededForInterval 1`] = `
Object {
  "components": Array [
    Object {
      "baselineCpu": 2,
      "baselineMemory": 10,
      "needCpu": 204,
      "needCpuForRequests": 200,
      "needMemory": 35,
      "needMemoryForRequests": 15,
      "needReplica": 2,
    },
  ],
  "needCpu": 204,
  "needMemory": 35,
}
`;

exports[`calculator calculateResourceNeededForRequestsInterval 1`] = `
Object {
  "components": Array [
    Object {
      "needCpuForRequests": 300,
      "needMemoryForRequests": 400,
      "requestToCpu": 3,
      "requestToMemory": 4,
    },
  ],
  "requests": 100,
}
`;
